Add reduced workflow for B+, B0s -> J/Psi X analyses

fchinu opened the pull request:
This PR adds the reduced workflows for the B+ and B0s meson analyses for their decays into a J/Psi and a K+ and phi, respectively. The considered decay of the J/Psi is that into a muon pair at midrapidity, and this workflow is meant to be used for the newly-developed HF software triggers for 2025 pp data taking. A `runJPsiToee` is also available in order to validate the workflow using the already existing MC simulations of B -> J/Psi X -> ee X.
In a following PR the workflow of B0 -> J/psi K*(892)0 -> (µµ) (Kpi) will be added, following what was done for the B0s -> J/Psi phi -> (µµ)(KK) workflow.

fchinu commented:
I would not feel super comfortable in fixing the O2 linter error
`error: Workflow name task-omegac0-to-omega-pi does not match its file name taskOmegac0ToOmegapi.cxx`, but of course if would you like me to do something in this PR I can do it, if the Omegac analysers are okay with that!

vkucera commented:
> I would not feel super comfortable in fixing the O2 linter error `error: Workflow name task-omegac0-to-omega-pi does not match its file name taskOmegac0ToOmegapi.cxx`, but of course if would you like me to do something in this PR I can do it, if the Omegac analysers are okay with that!

That's OK. That error is unrelated to your code and PR. Thanks for fixing the rest.

vkucera commented on PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:
```diff
@@ -0,0 +1,1047 @@
+        }
+      } // kaon loop
+      if (fillHfCandJPsi) { // fill JPsi table only once per JPsi candidate
+        double jPsiMass{0.f};
```
```suggestion
        double invMassJpsi{0.};
```
Use the recommended naming scheme.
`0.f` is a `float` not a `double`.

fchinu commented on PWGHF/Core/HfHelper.h:
```diff
@@ -871,6 +883,106 @@ class HfHelper
+  /// \param cuts B+ candidate selection per pT bin
+  /// \param binsPt pT bin limits
+  /// \return true if candidate passes all selections
+  template <typename T1, typename T2, typename T3>
+  bool selectionBplusToJPsiKTopol(const T1& candBp, const T2& cuts, const T3& binsPt)
+  {
```
The analysis we would like to perform will be done using the reduced workflow. I have put this function in the `HfHelper` to be consistent with what was done for the other B mesons, and this could be used in case someone needs to develop non-reduced tasks

vkucera commented on Common/Core/RecoDecay.h:
```diff
@@ -285,6 +286,18 @@ struct RecoDecay {
     return static_cast<double>(length) * static_cast<double>(mass) / p(mom);
   }
 
+  /// Calculates pseudoproper decay length.
+  /// \param posPV  {x, y, z} or {x, y} position of the primary vertex
+  /// \param posSV  {x, y, z} or {x, y} position of the secondary vertex
+  /// \param mom  {x, y, z} or {x, y} momentum array
+  /// \return pseudoproper decay length
+  template <std::size_t N, typename T, typename U, typename V, typename M = double>
+  static double pseudoProprDL(const T& posPV, const U& posSV, const std::array<std::array<V, 3>, N>& mom, const std::array<M, N> mass)
+  {
+    // t_xy = l_xy * m / pT
+    return distanceXY(posPV, posSV) * m(mom, mass) / std::apply([](const auto&... args) { return pt(args...); }, mom);
+  }
+
```
Why can you not use `ct`?

fchinu replied:
It's a slightly different variable, as with the pseudoproper decay length we can use the better resolution in the xy plane. It was also used by several papers so I would like to add it to our analysis as well

fchinu replied:
Ok, I renamed it, thank you!

vkucera reviewed (changes requested):
Hi @fchinu , thanks for updating the `RecoDecay` function. Please see my further comments.

vkucera commented on Common/Core/RecoDecay.h:
```diff
@@ -285,6 +286,18 @@ struct RecoDecay {
+  /// Calculates ctXY, the pseudoproper decay length.
+  /// \param posPV  {x, y, z} or {x, y} position of the primary vertex
+  /// \param posSV  {x, y, z} or {x, y} position of the secondary vertex
+  /// \param mom  {x, y, z} or {x, y} momentum array
```
Isn't `mom` a pack of 3-(or 2-)momentum arrays rather than a momentum array?

vkucera replied:
Btw, the current implementation does not seem to support {x, y} arrays.

alibuild commented:
Error while checking build/O2Physics/o2 for 24a927333d6da79cc8c323e92380f248e3c0c1cf at 2025-06-03 23:48:
```
## sw/BUILD/O2Physics-latest/log
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:112:84: error: 'hf_cuts_single_track' was not declared in this scope
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:112:84: error: no matching function for call to 'std::vector<double>::vector(<brace-enclosed initializer list>)'
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:112:83: error: expected primary-expression before '{' token
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:112:83: error: expected '}' before '{' token
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:112:83: error: could not convert '{"binsPtTrack", <expression error>}' from '<brace-enclosed initializer list>' to 'o2::framework::Configurable<std::vector<double> >'
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:112:83: error: expected ';' before '{' token
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:113:68: error: 'hf_cuts_single_track' has not been declared
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:113:104: error: 'hf_cuts_single_track' has not been declared
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:113:140: error: 'hf_cuts_single_track' has not been declared
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:113:177: error: 'hf_cuts_single_track' has not been declared
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:113:214: error: 'hf_cuts_single_track' has not been declared
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:113:311: error: could not convert '{"cutsTrackDCA", {<expression error>, <expression error>, <expression error>, <expression error>, <expression error>}, "Single-track selections per pT bin for bachelor track"}' from '<brace-enclosed initializer list>' to 'o2::framework::Configurable<o2::framework::LabeledArray<double> >'
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:115:74: error: 'hf_cuts_jpsi_to_mu_mu' was not declared in this scope
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:115:74: error: no matching function for call to 'std::vector<double>::vector(<brace-enclosed initializer list>)'
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:115:73: error: expected primary-expression before '{' token
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:115:73: error: expected '}' before '{' token
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:115:73: error: could not convert '{"binsPt", <expression error>}' from '<brace-enclosed initializer list>' to 'o2::framework::Configurable<std::vector<double> >'
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:115:73: error: expected ';' before '{' token
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:116:52: error: 'hf_cuts_jpsi_to_mu_mu' has not been declared
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:116:84: error: 'hf_cuts_jpsi_to_mu_mu' has not been declared
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:116:116: error: 'hf_cuts_jpsi_to_mu_mu' has not been declared
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:116:149: error: 'hf_cuts_jpsi_to_mu_mu' has not been declared
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:116:182: error: 'hf_cuts_jpsi_to_mu_mu' has not been declared
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:116:258: error: could not convert '{"cuts", {<expression error>, <expression error>, <expression error>, <expression error>, <expression error>}, "J/Psi candidate selection per pT bin"}' from '<brace-enclosed initializer list>' to 'o2::framework::Configurable<o2::framework::LabeledArray<double> >'
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:717:26: error: 'array' is not a member of 'o2::gpu::gpustd'; did you mean 'std::array'?
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:717:32: error: expected primary-expression before 'float'
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:720:156: error: 'dcaBach' was not declared in this scope
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:725:58: error: 'dcaBach' was not declared in this scope
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:804:30: error: 'array' is not a member of 'o2::gpu::gpustd'; did you mean 'std::array'?
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:804:36: error: expected primary-expression before 'float'
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:807:161: error: 'dcaBach2' was not declared in this scope
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:812:64: error: 'dcaBach2' was not declared in this scope
/sw/slc9_x86-64/O2/daily-20250603-0000-local2/include/Framework/StructToTuple.h:141:38: error: static assertion failed: You need to make sure that you have implicit constructors or that you call the explicit constructor correctly.
/sw/slc9_x86-64/GCC-Toolchain/v14.2.0-alice2-1/include/c++/14.2.0/bits/stl_construct.h:115:28: error: no matching function for call to 'construct_at(HfDataCreatorJPsiHadReduced*&)'
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/O2Physics/11368/24a927333d6da79cc8c323e92380f248e3c0c1cf/build_O2Physics_o2/pretty.html).

alibuild commented:
Error while checking build/O2Physics/o2 for ce8ce24c9df0e16ec39c0af7856c4892e4cb947f at 2025-06-04 15:20:
```
## sw/BUILD/O2Physics-latest/log
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:697:26: error: 'array' is not a member of 'o2::gpu::gpustd'; did you mean 'std::array'?
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:697:32: error: expected primary-expression before 'float'
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:700:156: error: 'dcaBach' was not declared in this scope
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:705:58: error: 'dcaBach' was not declared in this scope
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:784:30: error: 'array' is not a member of 'o2::gpu::gpustd'; did you mean 'std::array'?
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:784:36: error: expected primary-expression before 'float'
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:787:161: error: 'dcaBach2' was not declared in this scope
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:792:64: error: 'dcaBach2' was not declared in this scope
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/O2Physics/11368/ce8ce24c9df0e16ec39c0af7856c4892e4cb947f/build_O2Physics_o2/pretty.html).

alibuild commented:
Error while checking build/O2Physics/o2 for 80e8db66cac389602f62fa740fbe5c2ceb1473eb at 2025-06-04 15:37:
```
## sw/BUILD/O2Physics-latest/log
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:697:26: error: 'array' is not a member of 'o2::gpu::gpustd'; did you mean 'std::array'?
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:697:32: error: expected primary-expression before 'float'
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:700:156: error: 'dcaBach' was not declared in this scope
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:705:58: error: 'dcaBach' was not declared in this scope
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:784:30: error: 'array' is not a member of 'o2::gpu::gpustd'; did you mean 'std::array'?
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:784:36: error: expected primary-expression before 'float'
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:787:161: error: 'dcaBach2' was not declared in this scope
/sw/SOURCES/O2Physics/11368-slc9_x86-64/0/PWGHF/D2H/TableProducer/dataCreatorJPsiHadReduced.cxx:792:64: error: 'dcaBach2' was not declared in this scope
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/O2Physics/11368/80e8db66cac389602f62fa740fbe5c2ceb1473eb/build_O2Physics_o2/pretty.html).

fchinu commented:
Hi @vkucera @fgrosa, thanks for your comments! It seems to be building now, I should have addressed all your comments, please let me know in case I missed any

fgrosa commented:
Hi @fchinu the PR is good for me now, if @vkucera has no more comments we can merge it. Thanks!

vkucera commented on PWGHF/D2H/Utils/utilsRedDataFormat.h:
```diff
@@ -81,6 +81,30 @@ float getTpcTofNSigmaPi1(const T1& prong1)
 }
+
+/// Helper function to retrive PID information of bachelor kaon from b-hadron decay
+/// \param prong1 kaon track from reduced data format, aod::HfRedBachProng0Tracks
+template <typename T1>
```
Please add `\return` documentation.

fchinu commented:
Hi @vkucera! Thanks for your comments! I have implemented them (I had to rebase my branch to get your latest additions for the decay header, so this caused the commits to be added all at the end)

vkucera commented on PWGHF/Core/DecayChannels.h:
```diff
@@ -152,6 +154,7 @@ enum DecayChannelMain : int8_t {
   BplusToD0PiPi0,   // anti-D0 π+ π0
   BplusToD0PiGamma, // anti-D0 π+ γ0
   BplusToD0K,       // anti-D0 K+
+  BplusToJPsiK,     // J/ψ K+
```
```suggestion
  BplusToJpsiK,     // J/ψ K+
```

vkucera commented on PWGHF/Core/HfHelper.h:
```diff
@@ -173,6 +173,12 @@ class HfHelper
   }
 
+  template <typename T>
+  auto invMassBplusToJPsiK(const T& candidate)
```
```suggestion
  auto invMassBplusToJpsiK(const T& candidate)
```

vkucera commented on PWGHF/Core/HfHelper.h:
```diff
@@ -873,6 +885,106 @@ class HfHelper
+    auto candJPsi = candBp.jPsi();
+    float pseudoPropDecLen = candBp.decayLengthXY() * mcandBp / ptcandBp;
+
+    int pTBin = o2::analysis::findBin(binsPt, ptcandBp);
```
```suggestion
    int binPt = o2::analysis::findBin(binsPt, ptcandBp);
```

vkucera commented on PWGHF/Core/HfHelper.h:
```diff
@@ -873,6 +885,106 @@ class HfHelper
+  template <typename T1, typename T2, typename T3>
+  bool selectionBplusToJPsiKTopol(const T1& candBp, const T2& cuts, const T3& binsPt)
+  {
+    auto ptcandBp = candBp.pt();
+    auto mcandBp = invMassBplusToJPsiK(candBp);
+    auto ptJPsi = RecoDecay::pt(candBp.pxProng0(), candBp.pyProng0());
+    auto ptKa = RecoDecay::pt(candBp.pxProng1(), candBp.pyProng1());
+    auto candJPsi = candBp.jPsi();
```
```suggestion
    auto ptCandBp = candBp.pt();
    auto mCandBp = invMassBplusToJPsiK(candBp);
    auto ptJpsi = RecoDecay::pt(candBp.pxProng0(), candBp.pyProng0());
    auto ptKa = RecoDecay::pt(candBp.pxProng1(), candBp.pyProng1());
    auto candJpsi = candBp.jPsi();
```

vkucera commented on PWGHF/Core/HfHelper.h:
```diff
@@ -960,6 +1072,114 @@ class HfHelper
   }
 
+  // Apply topological cuts as defined in SelectorCuts.h
+  /// \param candBs Bs candidate
+  /// \param cuts Bs candidate selection per pT bin
+  /// \param binsPt pT bin limits
```
Add missing documentation.

vkucera commented on PWGHF/Core/HfHelper.h:
```diff
@@ -960,6 +1072,114 @@ class HfHelper
+  /// \param pidTrackKa PID status of trackKa (prong1 of B+ candidate)
+  /// \param acceptPIDNotApplicable switch to accept Status::NotApplicable
+  /// \return true if prong1 of B+ candidate passes all selections
+  template <typename T1 = int, typename T2 = bool>
+  bool selectionBsToJPsiPhiPid(const T1& pidTrackKa, const T2& acceptPIDNotApplicable)
+  {
+    if (!acceptPIDNotApplicable && pidTrackKa != TrackSelectorPID::Accepted) {
+      return false;
```
Why is the type of `pidTrackKa` templated?

fchinu commented:
Hi @vkucera, I have implemented the comments. You can check the  [Implement Vit's comments](https://github.com/AliceO2Group/O2Physics/pull/11368/commits/a65ce5e3cf7618a15f3373cf51d6eb159f149b2b) commit for the diff, as in the other commit I renamed the files and the diff is more difficult to retreive.

vkucera commented:
It seems that the MC matching is still using local enums and bitmaps instead of the enums from the main header.

fgrosa commented:
Hi @vkucera @fchinu sorry to comment only now, but the flags in the main header are not good as they are; all the flags for the channels with JPsi should be in the same enum (because one can be background for the other), and they should not be split between resonant and non resonant (unless for the Bs -> JPsi Phi you add a correspondent non resonant channel Bs -> JPsi KK, but I am not sure if it is necessary, since only the resonant should be relevant)

vkucera commented on PWGHF/Core/DecayChannels.h:
```diff
@@ -174,6 +174,23 @@ enum DecayChannelResonant : int8_t {
 };
+/// @brief beauty candidates: beauty to J/ψ decay channels
+enum DecayChannelToJpsiMain : int8_t {
+  BplusToJpsiK = 1, // J/ψ K+
+  B0ToJpsiPiK,      // J/ψ π- K+
+  BsToJpsiKK,       // J/ψ K+ K-
+  BcToJPsiPi,       // J/ψ π+
+  LbToJpsipK,       // J/ψ p K-
```
```suggestion
  // B+
  BplusToJpsiK = 1, // J/ψ K+
  // B0
  B0ToJpsiPiK,      // J/ψ π- K+
  // Bs0
  BsToJpsiKK,       // J/ψ K+ K-
  // Bc+
  BcToJPsiPi,       // J/ψ π+
  // Λb0
  LbToJpsipK,       // J/ψ p K-
```

vkucera commented on PWGHF/D2H/TableProducer/candidateCreatorBToJpsiReduced.cxx:
```diff
@@ -0,0 +1,465 @@
+  float myInvMassWindowJpsiK{1.}, myInvMassWindowJpsiPhi{1.}; // variable that will store the value of invMassWindowJpsiK (defined in dataCreatorJpsiKReduced.cxx)
+  double massBplus{0.}, massBs{0.};
+  double bz{0.};
+  o2::vertexing::DCAFitterN<2> df2; // fitter for B vertex (2-prong vertex fitter)
+  o2::vertexing::DCAFitterN<3> df3; // fitter for B vertex (2-prong vertex fitter)
+  o2::vertexing::DCAFitterN<4> df4; // fitter for B vertex (2-prong vertex fitter)
```
```suggestion
  o2::vertexing::DCAFitterN<2> df2; // fitter for B vertex (2-prong vertex fitter)
  o2::vertexing::DCAFitterN<3> df3; // fitter for B vertex (3-prong vertex fitter)
  o2::vertexing::DCAFitterN<4> df4; // fitter for B vertex (4-prong vertex fitter)
```

vkucera commented on PWGHF/D2H/TableProducer/candidateCreatorBToJpsiReduced.cxx:
```diff
@@ -0,0 +1,465 @@
+                            Cands const& candsJpsiThisColl,
+                            TTracks0 const& tracksLfDau0ThisCollision,
+                            TTracks1 const& tracksLfDau1ThisCollision,
+                            const float& invMass2JpsiHadMin,
+                            const float& invMass2JpsiHadMax)
```
Why do you pass these by reference?

vkucera reviewed (changes requested):
Please also make sure to apply IWYU and sort the includes with the latest clang-format configuration.
Also please fix the linter errors related to your code.

fchinu commented:
@vkucera `MegaLinter` is failing but the headers are ordered as described in the documentation (i.e. C++ headers at the end). The latest `.clang-format` was used to get this ordering. Can we merge as it is?

vkucera commented on Common/Core/RecoDecay.h:
```diff
@@ -23,6 +23,7 @@
 #include <cmath>     // std::abs, std::sqrt
 #include <cstdio>
 #include <utility> // std::move
+#include <tuple>   // std::apply
```
`std::apply` is directly used in the code.
Why should the relevant C++ header be included transitively and rely on the framework internals?

ktf replied:
For some reason I could not find the use of std::apply in the header.
